akka-cluster-sharding/src/main/scala/akka/cluster/sharding/internal/AbstractLeastShardAllocationStrategy.scala

```diff
@@ -0,0 +1,114 @@
+import akka.actor.ActorSystem
+import akka.actor.Address
+import akka.annotation.InternalApi
+import akka.cluster.Cluster
+import akka.cluster.ClusterEvent.CurrentClusterState
+import akka.cluster.Member
+import akka.cluster.MemberStatus
+import akka.cluster.sharding.ShardCoordinator.ActorSystemDependentAllocationStrategy
+import akka.cluster.sharding.ShardRegion.ShardId
+
+import scala.collection.immutable
+
+/**
+ * Common logic for the least shard allocation strategy implementations
+ *
+ * INTERNAL API
+ */
+@InternalApi
+private[akka] object AbstractLeastShardAllocationStrategy {
+  import MemberStatus._
+  private val JoiningCluster: Set[MemberStatus] = Set(Joining, WeaklyUp)
+  private val LeavingClusterStatuses: Set[MemberStatus] = Set(Leaving, Exiting, Down)
+
+  type AllocationMap = Map[ActorRef, immutable.IndexedSeq[ShardId]]
+
+  final case class RegionEntry(region: ActorRef, member: Member, shardIds: immutable.IndexedSeq[ShardId])
+
+  implicit object ShardSuitabilityOrdering extends Ordering[RegionEntry] {
+    override def compare(x: RegionEntry, y: RegionEntry): Int = {
+      val RegionEntry(_, memberX, allocatedShardsX) = x
+      val RegionEntry(_, memberY, allocatedShardsY) = y
+      if (memberX.status != memberY.status) {
+        // prefer allocating to nodes that are not on their way out of the cluster
+        val xIsLeaving = LeavingClusterStatuses(memberX.status)
+        val yIsLeaving = LeavingClusterStatuses(memberY.status)
+        JBoolean.compare(xIsLeaving, yIsLeaving)
+      } else if (memberX.appVersion != memberY.appVersion) {
+        // prefer nodes with the highest rolling update app version
+        memberY.appVersion.compareTo(memberX.appVersion)
+      } else {
+        // prefer the node with the least allocated shards
+        JInteger.compare(allocatedShardsX.size, allocatedShardsY.size)
+      }
+    }
+  }
+}
+
+/**
+
+ *
+ * INTERNAL API
+ */
+@InternalApi
+private[akka] abstract class AbstractLeastShardAllocationStrategy extends ActorSystemDependentAllocationStrategy {
+  import AbstractLeastShardAllocationStrategy._
+
+  @volatile private var cluster: Cluster = _
+
+  override def start(system: ActorSystem): Unit = {
+    cluster = Cluster(system)
+  }
+
+  // protected for testability
+  protected def clusterState: CurrentClusterState = cluster.state
+  protected def selfMember: Member = cluster.selfMember
+
+  final protected def isAGoodTimeToRebalance(regionEntries: Iterable[RegionEntry]): Boolean = {
+    // this will filter out member with no shard regions (bc role or not yet completed joining)
+    val shardRegionMembers = regionEntries.map { case RegionEntry(_, member, _) => member }
+    // avoid rebalance when rolling update is in progress
+    def allNodesSameVersion = shardRegionMembers.map(_.appVersion).toSet.size == 1
+    // rebalance requires ack from all anyway
+    def allRegionsReachable = clusterState.unreachable.diff(shardRegionMembers.toSet).isEmpty
```

Comment:
Hmmm. Same dc sounds reasonable but: if we skip when unreachable in dc and it is multi dc and there are unreachable in other dc rebalance won't be able to progress anyway until we fix #29589?

Reply:
I went with same dc since it should improve on things compared to current for non-multi-dc and does not make things worse for multi-dc, also added a note about #29598 there.

Reply:
sounds good, I added another note in https://github.com/akka/akka/issues/29589#issuecomment-692720031